tests/rados/test_pool_osd_recovery.py

```diff
@@ -0,0 +1,337 @@
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        log.debug(
+            "PG's are active + clean post OSD reboot, proceeding to restart next OSD"
+        )
+
+    log.info("All the planned primary OSD reboots have completed")
+    # Checking cluster health after the tests
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info("----Completed workflows 1. Reboot of single osd and health check----")
+
+    log.info("---- Starting workflow ----\n---- 2. Rolling Reboot OSD hosts ----")
+    for nodes in osd_nodes:
+        log.info(f"Rebooting node : {nodes.hostname}")
+        nodes.exec_command(sudo=True, cmd="reboot", long_running=True)
+
+        # Sleeping for 10 seconds and starting verification.
+        time.sleep(10)
+
+        # Waiting for recovery to post OSD host reboot
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        log.info(f"PG's are active + clean post reboot of host {nodes.hostname}")
+
+    log.info("Completed reboot of all the OSD hosts, Checking cluster health status")
+    # Checking cluster health after the test
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info(
+        "---- Completed workflows 2. Rolling Reboot OSD hosts and health check ----"
+    )
+
+    log.info(
+        "---- Starting workflow ----\n---- 3. Stopping and starting OSD daemons ----"
+    )
+    for pool in pools:
+        target_osd = rados_obj.get_pg_acting_set(pool_name=pool)[0]
+        log.debug(f"Stopping OSD : {target_osd} and checking health status")
+        if not rados_obj.change_osd_state(action="stop", target=target_osd):
+            log.error(f"Unable to stop the OSD : {target_osd}")
+            raise Exception("Execution error")
+
+        # Waiting for recovery to post OSD stop
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        log.debug(
+            f"PG's are active + clean post OSD stop of {target_osd}, proceeding to start OSD"
+        )
+
+        log.debug(f"Starting OSD : {target_osd} and checking health status")
+        if not rados_obj.change_osd_state(action="start", target=target_osd):
+            log.error(f"Unable to start the OSD : {target_osd}")
+            raise Exception("Execution error")
+
+        # Waiting for recovery to post OSD start
+        method_should_succeed(wait_for_clean_pg_sets, rados_obj)
+        log.debug(
+            f"PG's are active + clean post OSD start of {target_osd}, proceeding to restart next OSD"
+        )
+
+    log.info("Completed start and stop for all targeted OSDs. Checking cluster health")
+    # Checking cluster health after the tests
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info("--- Completed workflows 3. Stopping and starting OSD daemons ---")
+
+    log.info(
+        "---- Starting workflow ----\n---- 4. restart all OSD daemons belonging to single pg----"
+    )
+    for pool in pools:
```

Comment:
This scenario and  scenario 1. Reboot of single osd and health check are looks similar. Any additional verification are we doing here?

Reply:
No additional checks. The above scenario asked for only primary, then later all the PGs. The same has been followed.

Reply:
Hey Bharath, these are the scenarios :
1. Reboot of single osd (Primary PG OSD) and health check.
2. Rolling Reboot OSD hosts ( All OSDs belonging to 1 sample PG in the pool ).
3. Stopping and starting OSD daemons (Primary PG OSD) .
4. restart all OSD daemons belonging to single pg ( All OSDs belonging to 1 sample PG in the pool ) -> Difference from scenario 2 is that entire host is rebooted, here in 4, only the OSD service is rebooted.
5. Removal and addition of OSD daemons.
6. Removal and addition of OSD Hosts.

Reply:
LGTM